src/main/java/com/ust/sourcecourse/configuration/service/ColumnsService.java

```diff
@@ -0,0 +1,142 @@
+		if (optional.isPresent()) {
+			ProjectGroup projectGroup = optional.get();
+			return getColumnResponses(projectGroup);
+
+		} else {
+			throw new ResponseStatusException(HttpStatus.NOT_FOUND, "Project Group with uid " + groupId + " not found");
+		}
+
+	}
+
+	@Transactional
+	public List<ColumnsResponse> saveData(Long groupId, List<ColumnsRequest> columnsRequests) {
+		Optional<ProjectGroup> optional = projectGroupRepository.findById(groupId);
+
+		if (optional.isPresent()) {
+			ProjectGroup projectGroup = optional.get();
+			List<GroupColumn> groupColumns = projectGroup.getGroupColumns();
+			if (groupColumns == null) {
+				groupColumns = new ArrayList<>();
+			}
+
+			for (ColumnsRequest columnsRequest : columnsRequests) {
+				SourceColumn sourceColumn = columnsRepository.findById(columnsRequest.getSourceColumnUid())
+						.orElseThrow();
+
+				GroupColumn groupColumn = GroupColumn.builder().name(columnsRequest.getName())
+						.notes(columnsRequest.getNotes()).type(columnsRequest.getType())
+						.isPrimary(columnsRequest.isPrimary()).defaultValue(columnsRequest.getDefaultvalue())
+						.prefix(columnsRequest.getPreffix()).suffix(columnsRequest.getSuffix())
+						.projectGroup(projectGroup).sourceColumn(sourceColumn).build();
+
+				groupColumns.add(groupColumn);
+				projectGroup.setGroupColumns(groupColumns);
+				projectGroup = projectGroupRepository.save(projectGroup);
+			}
+			return getColumnResponses(projectGroup);
+
+		} else {
+			throw new ResponseStatusException(HttpStatus.NOT_FOUND, "Project Group with uid " + groupId + " not found");
+		}
+
+	}
+
+	private List<ColumnsResponse> getColumnResponses(ProjectGroup projectGroup) {
+		List<ColumnsResponse> columnsResponses = new ArrayList<>();
+		for (GroupColumn groupColumn : projectGroup.getGroupColumns()) {
+			AttributesInfo attributesInfo = new AttributesInfo();
+			attributesInfo.setName(groupColumn.getName());
+			attributesInfo.setNotes(groupColumn.getNotes());
+			attributesInfo.setType(groupColumn.getType());
+			attributesInfo.setPrimary(groupColumn.isPrimary());
+			attributesInfo.setDefaultValue(groupColumn.getDefaultValue());
+			attributesInfo.setPrefix(groupColumn.getPrefix());
+			attributesInfo.setSuffix(groupColumn.getSuffix());
+			SourceInfo sourceInfo = new SourceInfo(groupColumn.getSourceColumn().getName(),
+					groupColumn.getSourceColumn().getUid());
+			ColumnsResponse columnsResponse = new ColumnsResponse(groupColumn.getUid(), attributesInfo, sourceInfo);
+			columnsResponses.add(columnsResponse);
+		}
+		return columnsResponses;
+	}
+
+	public ColumnsResponse updateColumn(Long groupId, Long columnId, ColumnsRequest columnsRequest) {
+		Optional<ProjectGroup> optional = projectGroupRepository.findById(groupId);
+		if (optional.isPresent()) {
+	        ProjectGroup projectGroup = optional.get();
+	        List<GroupColumn> groupColumns = projectGroup.getGroupColumns();
+	        for(GroupColumn groupColumn : groupColumns)
```

Comment:
Instead of looping through, using groupColumnRepository.findById column Id can be passed and get the GroupColumn object directly. No need of group id here.

Reply:
resolved